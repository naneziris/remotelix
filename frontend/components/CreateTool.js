import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import Error from './ErrorMessage';

const CREATE_TOOL_MUTATION = gql`
  mutation CREATE_TOOL_MUTATION(
    $title: String!
    $description: String!
    $image: String
    $largeImage: String
    $url: String!
  ) {
    createTool(
      title: $title
      description: $description
      image: $image
      largeImage: $largeImage
      url: $url
    ) {
      id
    }
  }
`;

class CreateTool extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    url: '',
  };
  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'remotelix');

    const res = await fetch('https://api.cloudinary.com/v1_1/remotelix/image/upload', {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  };
  render() {
    return (
      <Mutation mutation={CREATE_TOOL_MUTATION} variables={this.state}>
        {(createTool, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              // Stop the form from submitting
              e.preventDefault();
              // call the mutation
              const res = await createTool();
              // change them to the single tool page
              console.log(res);
              Router.push({
                pathname: '/tool',
                query: { id: res.data.createTool.id },
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
                {this.state.image && (
                  <img width="200" src={this.state.image} alt="Upload Preview" />
                )}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter A Description"
                  required
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="url">
                Url
                <input
                  type="text"
                  id="url"
                  name="url"
                  placeholder="Url"
                  required
                  value={this.state.url}
                  onChange={this.handleChange}
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateTool;
export { CREATE_TOOL_MUTATION };
