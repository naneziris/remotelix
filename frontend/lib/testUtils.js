import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const fakeTool = () => ({
  __typename: 'Tool',
  id: 'abc123',
  url: 'www.remotelix.com/tool/nikos',
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  titleToLowerCase: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
  category: '',
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeTool,
  fakeUser,
};
