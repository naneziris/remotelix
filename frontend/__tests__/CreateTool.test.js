import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateTool, { CREATE_TOOL_MUTATION } from '../components/CreateTool';
import { fakeTool } from '../lib/testUtils';

const dogImage = 'https://dog.com/dog.jpg';

// mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [{ secure_url: dogImage }],
  }),
});

describe('<CreateTool/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateTool />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test="form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('uploads a file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateTool />
      </MockedProvider>
    );
    const input = wrapper.find('input[type="file"]');
    input.simulate('change', { target: { files: ['fakedog.jpg'] } });
    await wait();
    const component = wrapper.find('CreateTool').instance();
    expect(component.state.image).toEqual(dogImage);
    expect(component.state.largeImage).toEqual(dogImage);
    expect(global.fetch).toHaveBeenCalled();
    global.fetch.mockReset();
  });

  it('handles state updating', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateTool />
      </MockedProvider>
    );
    wrapper.find('#title').simulate('change', { target: { value: 'Testing', name: 'title' } });
    wrapper
      .find('#description')
      .simulate('change', { target: { value: 'This is a really nice tool', name: 'description' } });

    expect(wrapper.find('CreateTool').instance().state).toMatchObject({
      title: 'Testing',
      description: 'This is a really nice tool',
    });
  });
  it('creates an tool when the form is submitted', async () => {
    const tool = fakeTool();
    const mocks = [
      {
        request: {
          query: CREATE_TOOL_MUTATION,
          variables: {
            title: tool.title,
            description: tool.description,
            image: '',
            largeImage: '',
            url: '',
          },
        },
        result: {
          data: {
            createTool: {
              ...fakeTool,
              id: 'abc123',
              __typename: 'Tool',
            },
          },
        },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateTool />
      </MockedProvider>
    );
    // simulate someone filling out the form
    wrapper.find('#title').simulate('change', { target: { value: tool.title, name: 'title' } });
    wrapper
      .find('#description')
      .simulate('change', { target: { value: tool.description, name: 'description' } });
    // mock the router
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');
    await wait(50);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({ pathname: '/tool', query: { id: 'abc123' } });
  });
});
