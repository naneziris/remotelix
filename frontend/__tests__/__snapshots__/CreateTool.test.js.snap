// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateTool/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1x8epu1-0 fzKnXg"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="category"
    >
      Category
      <input
        id="category"
        name="category"
        onChange={[Function]}
        placeholder="Category"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter A Description"
        required={true}
        value=""
      />
    </label>
    <label
      htmlFor="url"
    >
      Url
      <input
        id="url"
        name="url"
        onChange={[Function]}
        placeholder="Url"
        required={true}
        type="text"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
